plugins/inputs/beat/beat.go

```diff
@@ -0,0 +1,278 @@
+		CollectLibbeatStats:  true,
+		CollectSystemStats:   true,
+		CollectFilebeatStats: true,
+		Method:               "GET",
+		Headers:              make(map[string]string),
+		HostHeader:           "",
+		Timeout:              internal.Duration{Duration: time.Second * 5},
+	}
+}
+
+func (beat *Beat) Description() string {
+	return description
+}
+
+func (beat *Beat) SampleConfig() string {
+	return sampleConfig
+}
+
+// createHttpClient create a clients to access API
+func (beat *Beat) createHttpClient() (*http.Client, error) {
+	tlsConfig, err := beat.ClientConfig.TLSConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+		Timeout: beat.Timeout.Duration,
+	}
+
+	return client, nil
+}
+
+// gatherJsonData query the data source and parse the response JSON
+func (beat *Beat) gatherJsonData(url string, value interface{}) error {
+
+	var method string
+	if beat.Method != "" {
+		method = beat.Method
+	} else {
+		method = "GET"
+	}
+
+	request, err := http.NewRequest(method, url, nil)
+	if err != nil {
+		return err
+	}
+
+	if (beat.Username != "") || (beat.Password != "") {
+		request.SetBasicAuth(beat.Username, beat.Password)
+	}
+	for header, value := range beat.Headers {
+		request.Header.Add(header, value)
+	}
+	if beat.HostHeader != "" {
+		request.Host = beat.HostHeader
+	}
+
+	response, err := beat.client.Do(request)
+	if err != nil {
+		return err
+	}
+
+	defer response.Body.Close()
+
+	err = json.NewDecoder(response.Body).Decode(value)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (beat *Beat) gatherInfoTags(url string) (map[string]string, error) {
+	beatInfo := &BeatInfo{}
+
+	err := beat.gatherJsonData(url, beatInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	tags := map[string]string{
+		"beat_beat":    beatInfo.Beat,
+		"beat_id":      beatInfo.UUID,
+		"beat_name":    beatInfo.Name,
+		"beat_host":    beatInfo.Hostname,
+		"beat_version": beatInfo.Version,
+	}
+
+	return tags, nil
+}
+
+func (beat *Beat) gatherStats(accumulator telegraf.Accumulator) error {
+	beatStats := &BeatStats{}
+
+	infoUrl, err := url.Parse(beat.URL + suffixInfo)
+	if err != nil {
+		return err
+	}
+	statsUrl, err := url.Parse(beat.URL + suffixStats)
+	if err != nil {
+		return err
+	}
+
+	tags, err := beat.gatherInfoTags(infoUrl.String())
+	if err != nil {
+		return err
+	}
+
+	err = beat.gatherJsonData(statsUrl.String(), beatStats)
+	if err != nil {
+		return err
+	}
+
+	if beat.CollectBeatStats {
+		flattenerBeat := jsonparser.JSONFlattener{}
+		err := flattenerBeat.FlattenJSON("", beatStats.Beat)
+		if err != nil {
+			return err
+		}
+		accumulator.AddFields("beat", flattenerBeat.Fields, tags)
+	}
+
+	if beat.CollectFilebeatStats {
+		flattenerBeat := jsonparser.JSONFlattener{}
+		err := flattenerBeat.FlattenJSON("", beatStats.FileBeat)
+		if err != nil {
+			return err
+		}
+		accumulator.AddFields("beat_filebeat", flattenerBeat.Fields, tags)
```

Comment:
Your metric name `beat_filebeat` does not follow the convention (default: plugin name). This is especially bad if people use `name_override` to customize the output and suddenly don't get the three distinct streams but an undistinguishible mixture.
How about adding the "information type" (`filebeat` in this case) as a tag (e.g. `source="filebeat"`)?

Reply:
my thinking here was that it made more sense to have separate measurements as there's not much overlap in the responses, it's the interface that is consistent between the endpoints.

don't really have a strong opinion on it though.

Reply:
Go ahead as-is for now. I have to discuss this more with Influx-people... Many plugins do this and if you don't be careful you easily screw up the metrics when using `name_override` as it doesn't care about the suffix...
Mark resolved please.